contrib/src/main/java/com/datatorrent/contrib/hbase/HBaseScanOperator.java

```diff
@@ -40,27 +51,107 @@
  * @tags hbase, scan, input operator
  * @since 0.3.2
  */
-public abstract class HBaseScanOperator<T> extends HBaseInputOperator<T>
+public abstract class HBaseScanOperator<T> extends HBaseInputOperator<T> implements Operator.ActivationListener<Context>
 {
+  public static final int DEF_HINT_SCAN_LOOKAHEAD = 2;
+  public static final int DEF_QUEUE_SIZE = 1000;
+  public static final int DEF_SLEEP_MILLIS = 10;
+
+  private String startRow;
+  private String endRow;
+  private String lastReadRow;
+  private int hintScanLookahead = DEF_HINT_SCAN_LOOKAHEAD;
+  private int queueSize = DEF_QUEUE_SIZE;
+  private int sleepMillis = DEF_SLEEP_MILLIS;
+  private Queue<Result> resultQueue;
+  private boolean threadFailed = false;
+
+  @AutoMetric
+  protected long tuplesRead;
+
+  // Transients
+  protected transient Scan scan;
+  protected transient ResultScanner scanner;
+  protected transient Thread readThread;
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    super.setup(context);
+    resultQueue = Queues.newLinkedBlockingQueue(queueSize);
+  }
+
+  @Override
+  public void activate(Context context)
+  {
+    startReadThread();
+  }
+
+  protected void startReadThread()
+  {
+    try {
+      scan = operationScan();
+      scanner = getStore().getTable().getScanner(scan);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    readThread = new Thread(new Runnable() {
+      @Override
+      public void run()
+      {
+        try {
+          Result result;
+          while ((result = scanner.next()) != null) {
+            while (!resultQueue.offer(result)) {
+              Thread.sleep(sleepMillis);
+            }
+          }
+        } catch (Exception e) {
+          logger.debug("Exception in fetching results {}", e.getMessage());
+          threadFailed = true;
+          throw new RuntimeException(e);
+        } finally {
+          scanner.close();
+        }
+      }
+    });
+    readThread.start();
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    super.beginWindow(windowId);
+    tuplesRead = 0;
+  }
 
   @Override
   public void emitTuples()
   {
+    if (!readThread.isAlive() && threadFailed) {
+      throw new RuntimeException("Exception in scan thread");
```

Comment:
I think if we should not just restart the thread without understanding what the issue was. In any case, this exception will ultimately kill the operator and restart it from a consistent state instead.